tests/test_stepper_motor.py

```diff
@@ -0,0 +1,53 @@
+"""Tests for DummyStepperMotor."""
+from itertools import chain
+
+import pytest
+
+from finesse.config import ANGLE_PRESETS
+from finesse.hardware.dummy_stepper_motor import DummyStepperMotor
+
+
+@pytest.mark.parametrize("steps", (len(ANGLE_PRESETS), 1000, 10000000))
+def test_constructor_good(steps: int) -> None:
+    """Check valid arguments to constructor."""
+    DummyStepperMotor(steps)
+
+
+@pytest.mark.parametrize("steps", (-1000000, -1000, *range(-1, len(ANGLE_PRESETS))))
+def test_constructor_bad(steps: int) -> None:
+    """Check invalid arguments to constructor."""
+    with pytest.raises(ValueError):
+        DummyStepperMotor(steps)
```

Comment:
The normal use when using parametric tests is that you provide as input to the test the input parameters and the expected outputs/outcomes. So, these two functions could be written in a more concise manner as (see [VirtualRainforest](https://github.com/ImperialCollegeLondon/virtual_rainforest/blob/89407edba9c707f33281dd9051fecbee7ec716dd/tests/test_models.py#L43) for a more complex example):

```python
from contextlib import nullcontext as does_not_raise


@pytest.mark.parametrize(
    "steps,raises",
    [
        [
            num,
            pytest.raises(ValueError) if num < len(ANGLE_PRESETS) else does_not_raise(),
        ]
        for num in range(-5, len(ANGLE_PRESETS) + 5)
    ],
)
def test_constructor(steps: int, raises: Any) -> None:
    """Check invalid arguments to constructor."""
    with raises:
        DummyStepperMotor(steps)
```

Another option is to put the logic to decide on the outcome inside the test:

```python
from contextlib import nullcontext as does_not_raise


@pytest.mark.parametrize(
    "steps",
    list(range(-5, len(ANGLE_PRESETS) + 5)),
)
def test_constructor(steps: int) -> None:
    """Check invalid arguments to constructor."""
    raises = (
        pytest.raises(ValueError) if steps < len(ANGLE_PRESETS) else does_not_raise()
    )
    with raises:
        DummyStepperMotor(steps)
```

Assuming `mypy` does not complain much, I think this approach is more concise than splitting into good and bad tests.

Reply:
Oh that's really cool. Thanks!